docs/source/notebooks/QLOQ_QUBO_tutorial.ipynb

```diff
@@ -0,0 +1,1216 @@
+    "        probabilities = list(output_dict.values())\n",
+    "        values = [expectation_value(np.array(list(state)).astype(int), qubo_matrix)\n",
+    "                  for state in output_dict.keys()]\n",
+    "\n",
+    "        loss = compute_cvar(probabilities, values, alpha)\n",
+    "\n",
+    "        # Track the best result\n",
+    "        bitstring = max(output_dict, key=output_dict.get)\n",
+    "        if best_result[0] is None or loss < best_result[0][0]:\n",
+    "            best_result[0] = (loss, bitstring)\n",
+    "\n",
+    "        iteration[0] += 1\n",
+    "        print(f\"Iteration {iteration[0]}: Loss = {loss}, Best bitstring = {bitstring}\")\n",
+    "\n",
+    "        return loss\n",
+    "\n",
+    "    return objective_function, best_result\n"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 10,
+   "metadata": {
+    "ExecuteTime": {
+     "end_time": "2025-01-16T10:21:45.918288800Z",
+     "start_time": "2025-01-16T10:21:45.868770600Z"
+    }
+   },
+   "outputs": [],
+   "source": [
+    "\n",
+    "\n",
+    "def optimize_qubo(qubo_matrix: np.ndarray,\n",
+    "                  input_state: str,\n",
+    "                  group_sizes: List[int],\n",
+    "                  layers: List[str],\n",
+    "                  sampling_size: int = 100000,\n",
+    "                  alpha: float = 0.5,\n",
+    "                  maxiter: int = 50) -> Tuple[float, str, np.ndarray]:\n",
+    "    \"\"\"\n",
+    "    Run the optimization using COBYLA to find phases that minimize CVaR for the given QUBO problem.\n",
+    "    - qubo_matrix: your QUBO cost matrix\n",
+    "    - input_state: initial bitstring\n",
+    "    - group_sizes: e.g. [2,2,2]\n",
+    "    - layers: e.g. [\"Y\"]\n",
+    "    - sampling_size: how many shots per circuit evaluation\n",
+    "    - alpha: fraction for CVaR\n",
+    "    - maxiter: max COBYLA iterations\n",
+    "\n",
+    "    Returns:\n",
+    "      final_loss (float): best CVaR found\n",
+    "      best_bitstring (str): best measured bitstring\n",
+    "      optimal_phases (np.ndarray): the phase vector that achieved the best result\n",
+    "    \"\"\"\n",
+    "    ansatz = catalog[\"QLOQ ansatz\"]\n",
+    " \n",
+    "    group_sizes_p = [Encoding.DUAL_RAIL if x == 1 else eval(f\"Encoding.QUDIT{x}\") for x in group_sizes]\n",
+    "    nb_phases = ansatz.get_parameter_nb(group_sizes_p, len(layers))\n",
+    "\n",
+    "    # Random initial guess for the phases\n",
+    "    initial_phases = np.random.uniform(0, 2*np.pi, nb_phases)\n",
+    "\n",
+    "    # Build the objective function\n",
+    "    objective_function, best_result = create_objective_function(\n",
+    "        qubo_matrix, input_state, group_sizes, layers, sampling_size, alpha\n",
+    "    )\n",
+    "\n",
+    "    # Define a custom inequality to keep phases in [0, 2π] if we want\n",
+    "    def constraints(x):\n",
```

Comment:
Comment this method if it is still not used (because commented at usage in `minimize`)